CLS Silverstone: enable the init sequence for all of the QSFPDD CMIS4…

ds952811 opened the pull request:
… modules

Signed-off-by: Dante Su <dante.su@broadcom.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
1. Enable the initialization sequence for all of the QSFPDD CMIS4 modules, which used to be limited to InnoLight only.
2. Stop issuing software reset to the QSFPDD DACs even though its module state is not ModuleReady for 3+ seconds.
3. Lower the clock rate of FPGA-I2C to have the best transceiver compatibility.

**- How I did it**
See What I did

**- How to verify it**
1. Have both InnoLight and Eoptolink CMIS4 modules attached to the DUT
2. Issue 'show interfaces transceiver eeprom' to verify if their application selection is '2'
3. Intentionally disable anyone of the 4 links of the transceiver, and make sure the other 3 links are still UP

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

ds952811 committed:
CLS Silverstone: enable the init sequence for all of the QSFPDD CMIS4 modules

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 commented:
**PLEASE DO NOT** approve this review until the Broadcom QA finished the internal tests, I'll later send an all-clear flag once it's done.

This PR is early access to review what are the proposed changes to address the CMIS4 issues, while the new test image has yet generated, the DUT with corresponding changes manually applied are verified, please refer to the attached log for details.
[PR95-ENG-TEST.log](https://github.com/zhenggen-xu/sonic-buildimage/files/5594432/PR95-ENG-TEST.log)

ds952811 commented on platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c:
```diff
@@ -1431,7 +1431,7 @@ static int smbus_access(struct i2c_adapter *adapter, u16 addr,
     }
 
-#if 1 /* 100 kHz */
+#if 0 /* 100 kHz */
```
Honestly, this is unlikely to be the root cause of I2C failure reported yesterday, however it's the only thing under our control without having the Celestica engaged, that's why I did this.
Please note it's always working on our test unit no matter if high-speed (BIT6-0x40) is enabled or not.

zhenggen-xu replied:
If we turn off the high speed mode,  how slow would it be?  I don't feel we should do this unless we feel it was necessary.

ds952811 replied:
1. We will have ~30% delays for each I2C operations when the BIT6 (high-speed) is cleared
2. Based on the latest test report from Broadcom QA, slowing down the clock does not improve the transceiver compatibility.
And hence, I'll revert this change.

Note: None of the QSFPDD optics is available at my side, so I'm counting on QA for the verifications

e.g.
Dante,

It appears Eoptolink 3.0 broke. It took ~17 minutes for link up after remove/insert
Please see the logs below

Regards,
-kevin

......
Nov 25 21:47:25.015491 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:47:25.015491 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:48:06.595530 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:48:06.595530 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:49:36.731519 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:49:36.731519 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:50:11.383533 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:50:11.383533 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:50:39.103534 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:50:39.103534 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:51:34.543512 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0

ds952811 commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -524,7 +551,14 @@ def get_transceiver_change_event(self, timeout=0):
+                    # its module state could get stuck at ModuleLowPwr upon reinsertion, and a software
+                    # reset is necessary to get it recovered.
+                    buf = self._read_eeprom_devid(x, self.IDENTITY_EEPROM_ADDR, 85, 1)
+                    if (buf is not None) and (int(buf[0], 16) != 0x03):
```
This is to address the DAC failure below:
pmon#sfputil.py: write failed: [Errno 110] Connection timed out

Some of the DAC (Passive Copper) do not support software reset, and hence such a write attempt will fail

zhenggen-xu replied:
In case of DAC, should we even add failure counter and come here,  See line 553:
`self.mod_failure[x] += 1`

zhenggen-xu replied:
another generic question,  it seems we are keep monitoring all the settings and do reset/init etc all the time.  It is a bit concerning if we that could be disruptive.  Thinking maybe we could only do these steps when we got the optical insert event,  and finish the bring up process once.  of course, we will do it again if insert happens again. Let me know.

ds952811 replied:
1. We'll actually do nothing in the case of DAC, so I'm not sure why we need to keep the counter advanced? Do we want to have any other automatic recovery for DAC?

2. The reason why this is not handled in the insertion event is that, the I2C read could fail and it was observed during the QA test last week. However, we could add a inited flag to make sure this is only performed once and only once, let me update the code logic

zhenggen-xu replied:
1- that is exactly my point, we should not advance the counter and then avoid the write here,  we should eliminate the counter advance in the code earlier so it won't even be a failure case.

ds952811 replied:
copy that, and thanks, I'll update the code accordingly

ds952811 commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -477,7 +496,15 @@ def _init_cmis_module(self, port_num):
+            rev = int(buf[1], 16)
+        if rev >= 0x40:
+            return self._init_cmis4_module(port_num, xcvr, self.hwsku)
+
```
All of the CMIS3 modules we have are working fine with 4x100G in its default application mode, and hence this init sequence will only be performed on CMIS4 modules.

Note: The definition of the datapath initialiation slightly differs from CMIS3 to CMIS4, and hence we can't use _init_cmis4_module() on CMIS3 modules.

ds952811 commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -450,7 +469,7 @@ def _init_cmis_module_custom(self, port_num, xcvr, hwsku):
         # Initialize datapath
         self._write_byte(port_num, self.IDENTITY_EEPROM_ADDR, 0x10, 128, 0x00)
-        time.sleep(0.5)
+        time.sleep(1)
```
The UT result shows that we need a few seconds for the actual configuration status correctly reflected on the register, but we don't really want to wait too long in the xcvrd, so I slightly enlarge the delay from 500ms to 1sec, hope this could reduce the redudant errors in the syslog

ds952811 commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -409,16 +409,32 @@ def _write_byte(self, port_num, devid, page, off, val):
-            return True
-
-        log_info("PORT {0}: {1}: _init_cmis_module_custom".format(port_num, xcvr))
+    def _get_application_code(self, port_num, host_intf):
```
There is no guarantee that 100G PAM4 mode is always the application 2, and hence we'll need to decode the application advertisement and figure the correct application code

ds952811 committed:
CLS Silverstone: ensure the cmis4 init will only be performed once, and revert the clock of FPGA-I2C to high-speed

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -437,7 +439,7 @@ def _init_cmis4_module(self, port_num, xcvr, hwsku):
 
         # Allow 1s for software reset
-        self._write_byte(port_num, self.IDENTITY_EEPROM_ADDR, -1, 26, 0x08)
+        self._write_byte(port_num, self.IDENTITY_EEPROM_ADDR, -1, 26, 0x58)
```
Ensure the module will be placed in LowPower mode

ds952811 commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -469,10 +471,16 @@ def _init_cmis4_module(self, port_num, xcvr, hwsku):
-        buf = self._read_eeprom_devid(port_num, self.IDENTITY_EEPROM_ADDR, self._page_to_flat(202, 0x11), 4)
-        err = "".join(buf)
+        err = '00000000'
+        for t in range(10):
```
Eoptolink CMIS4 took 6 seconds for config status to get populated to the register, and hence 10 sec should be sufficient

ds952811 commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -528,49 +536,63 @@ def get_transceiver_change_event(self, timeout=0):
                 if buf is None:
                     continue
                 err = "".join(buf)
                 if err != '11111111':
```
Just for a record, this checker is necessary to prevent redundant CMIS4 init upon warm-reboot, otherwise significant packet drops will be observed.

ds952811 commented on platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c:
```diff
@@ -1431,7 +1431,7 @@ static int smbus_access(struct i2c_adapter *adapter, u16 addr,
     }
 
-#if 0 /* 100 kHz */
+#if 1 /* 100 kHz */
```
Revert the clock to high-speed, unless it's unlikely to be the root cause, and we actually have this enabled in ICOS, and I never got any issues for this

zhenggen-xu commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -409,35 +411,54 @@ def _write_byte(self, port_num, devid, page, off, val):
+            return 1
+        tbl = self.parse_media_type(buf, 0)
+        app = 1
+        if tbl is not None:
```
This looks very similar to the code at line 656.  Can we make it as a function and reuse?

ds952811 replied:
It's done, please check it out

ds952811 committed:
CLS Silverstone: code refactoring for the application advertisement decoder

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 committed:
CLS Silverstone: sfputil.py: fix Eoptolink CMIS3 support

It's a known issue that Eoptolink CMIS3 requires a software reset after
re-insertion, and a burst of software reset messages observed when its
state != ModuleReady for 5 seconds.

And hence, we'll have the following changes in this commit

1. Update the checker from (state != ModuleReady) to (state == ModuleLowPwr),
   which is exactly the behavior observed on the Eoptolink CMIS3, and its state
   will become ModulePwrDn after reset.
2. Add 1-second delay to the software reset, and hence it's impossible to have
   a burst of reset at a time
3. Ensure the failure counter gets reset when its state != ModuleLowPwr

Signed-off-by: Dante Su <dante.su@broadcom.com>

zhenggen-xu commented:
As you mentioned, please send the all clear flag once the QA is done.

ds952811 commented:
> As you mentioned, please send the all clear flag once the QA is done.

Although the hotfix in the last commit 'CLS Silverstone: sfputil.py: fix Eoptolink CMIS3 support' has already verified by the QA, we still need to generate a new build with this change, and I'm working on it, will let you know once it's ready.

ds952811 commented:
> > As you mentioned, please send the all clear flag once the QA is done.
> 
> Although the hotfix in the last commit 'CLS Silverstone: sfputil.py: fix Eoptolink CMIS3 support' has already verified by the QA, we still need to generate a new build with this change, and I'm working on it, will let you know once it's ready.

After the internal discussion with my manager, Babu, it sounds like you're not using the binary from Broadcom, and you will use only the image rebuilt from the source package.

In this case, you don't need to wait for the new build(image) from my end, please approve and merge the changes if no any concern
